press/press/doctype/site/site.py

```diff
@@ -370,6 +378,68 @@ def validate_auto_update_fields(self):
 		# Validate day of month
 		if not (1 <= self.update_on_day_of_month <= 31):
 			frappe.throw("Day of the month must be between 1 and 31 (included)!")
+		# If site is on public bench, don't allow to disable auto updates
+		is_group_public = frappe.get_cached_value("Release Group", self.group, "public")
+		if self.skip_auto_updates and is_group_public:
+			frappe.throw("Auto updates can't be disabled for sites on public benches!")
+
+	def validate_site_plan(self):
+		if hasattr(self, "subscription_plan") and self.subscription_plan:
+			"""
+			If `release_groups` in site plan is empty, then site can be deployed in any release group.
+			Otherwise, site can only be deployed in the clusters mentioned in the release groups.
+			"""
+			release_groups = frappe.db.get_all(
+				"Site Plan Release Group",
+				pluck="release_group",
+				filters={
+					"parenttype": "Site Plan",
+					"parentfield": "release_groups",
+					"parent": self.subscription_plan,
+				},
+			)
+			clusters = frappe.db.get_all(
+				"Bench", pluck="cluster", filters={"group": ("in", release_groups)}
+			)
+			is_valid = len(clusters) == 0 or self.cluster in clusters
+			if not is_valid:
+				frappe.throw(
+					"In {0}, you can't deploy site in {1} cluster".format(
+						self.subscription_plan, self.cluster
+					)
+				)
+			"""
+			If `allowed_apps` in site plan is empty, then site can be deployed with any apps.
+			Otherwise, site can only be deployed with the apps mentioned in the site plan.
+			"""
+			allowed_apps = frappe.db.get_all(
+				"Site Plan Allowed App",
+				pluck="app",
+				filters={
+					"parenttype": "Site Plan",
+					"parentfield": "allowed_apps",
+					"parent": self.subscription_plan,
+				},
+			)
+			if allowed_apps:
+				selected_apps = list(
+					set(
+						frappe.db.get_all(
+							"Site App",
+							pluck="app",
+							filters={
+								"parenttype": "Site",
+								"parentfield": "apps",
+								"parent": self.name,
+							},
+						)
+					)
+				)
+				for app in selected_apps:
+					if app not in allowed_apps:
+						frappe.throw(
+							"In {0}, you can't deploy site with {1} app".format(self.subscription_plan, app)
+						)
 
```

Comment:
You might want to use an early exist here to make this more readable. Ref: https://gist.github.com/18alantom/d9f0565c0f42d6a71311d4a3093a1331#file-code_readability-md

Reply:
In the code, all the parts should execute based on different condition. We need to check for `allowed_apps` and `release_groups` independently. I didn't find any base case to do early exit.